api.py

```diff
@@ -0,0 +1,438 @@
+import os
+
+from flask import Flask, request
+
+from flask_restful import Resource, Api, reqparse, fields, marshal
+
+from flask_httpauth import HTTPBasicAuth
+
+from sqlalchemy import create_engine, MetaData, desc, func
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy.ext.serializer import loads, dumps
+
+from sqlalchemy_paginator import Paginator
+from sqlalchemy_paginator.exceptions import EmptyPage
+
+from models import models
+
+app = Flask(__name__)
+app.config.from_object(os.environ['APP_SETTINGS'])
+
+api = Api(app)
+
+auth = HTTPBasicAuth()
+
+def init_session():
+	"""
+	This method configures and the app's current database and returns a session instance for use
+	during CRUD operations.
+	"""
+	engine = create_engine(app.config.get('DATABASE_URL'))
+	session = sessionmaker()
+	session.configure(bind=engine)
+	return session()
+
+manager = init_session()
+
+access_denied = {'message': 'Access denied!'}
+
+def get_request_token():
+	"""
+	This method is used to retrieve a user's token from the username key of the request's header.
+	"""
+	return request.headers.get('username')
+
+def is_bucketlist_owner(bucketlist):
+	"""
+	This method requires a bucketlist as an argument and checks its creator against the currently
+	logged in user. If the current user has the same username as the created_by field of the bucketlist,
+	this method returns True, otherwise, it returns False.
+	"""
+	token = get_request_token()
+	# Ensure the owner od the bucketlist is the only one who can update it
+	if models.User.verify_auth_token(token, manager).username == bucketlist.created_by:
+		return True
+	else:
+		return False
+
+def paging(fields, paginator, page):
+	"""
+	This method receives field, paginator and page arguments. It uses paginator and page arguments
+	to paginate sqlalchemy query sets. The fields argument is used by marshal to return serialized results.
+	"""
+	try:
+		return marshal(paginator.page(page).object_list, fields), 200 
+	except EmptyPage:
+		return {'message': "Page doesn't exist"}, 404
+
+@auth.verify_password
+def verify_password(username_or_token, password):
+	"""
+	The two arguments are not required in this function's body. They are simply there to avoid errors
+	from the Flask-HTTPAuth library.
+	The method is improvised to authenticate users from tokens instead of authentication via sessions
+	as Flask-HTTPAuth originally intented.
+	"""
+	token = get_request_token()
+	if token:
+		user = models.User.verify_auth_token(token, manager)
+		if user:
+			return user
+		else:
+			return False
+	else:
```

Comment:
**Code Quality - Nitpick**
If the `else` block has only on statement (`return` expression), you can do without the `else` key word, thus:
```python
if user:
    return user
return False
```